shanoir-ng-datasets/src/main/java/org/shanoir/ng/dataset/service/CreateStatisticsService.java

```diff
@@ -42,46 +47,124 @@ private File recreateFile(final String fileName) throws IOException {
 
     @Async
     @Transactional
-    public void createStats(String studyNameInRegExp, String studyNameOutRegExp, String subjectNameInRegExp, String subjectNameOutRegExp, ShanoirEvent event, String params) throws RestServiceException, IOException {
+    public void createStats(String studyNameInRegExp, String studyNameOutRegExp, String subjectNameInRegExp, String subjectNameOutRegExp, ShanoirEvent event, String params) throws IOException {
         float progress = 0;
         String tmpDir = System.getProperty(JAVA_IO_TMPDIR);
         File userDir = DatasetFileUtils.getUserImportDir(tmpDir);
-        File statisticsFile = recreateFile(userDir + File.separator + "shanoirExportStatistics_" + event.getId() + ".tsv");
         File zipFile = recreateFile(userDir + File.separator + "shanoirExportStatistics_" + event.getId() + ZIP);
+        int startRow = 0;
+        int blocSize = 50000;
 
-        // Get the data
-        try (FileOutputStream fos = new FileOutputStream(statisticsFile);
-             BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(fos))){
+        event.setMessage("Querying size...");
+        eventService.publishEvent(event);
 
-            List<Object[]> results = datasetService.queryStatistics(studyNameInRegExp, studyNameOutRegExp, subjectNameInRegExp, subjectNameOutRegExp);
+        int procedureSize = querySize(studyNameInRegExp, studyNameOutRegExp, subjectNameInRegExp, subjectNameOutRegExp);
+        if (procedureSize > -1) {
 
-            for (Object[] or : results) {
+            if (procedureSize < blocSize) {
+                blocSize = procedureSize;
+            }
 
-                progress += 1f / results.size();
-                event.setProgress(progress);
+            // Get the data
+            try (FileOutputStream fos = new FileOutputStream(zipFile);
+                 ZipOutputStream zos = new ZipOutputStream(fos)) {
+
+                ZipEntry zipEntry = new ZipEntry("shanoirExportStatistics_" + event.getId() + ".tsv");
+
+                zos.putNextEntry(zipEntry);
+                OutputStreamWriter outputStreamWriter = new OutputStreamWriter(zos);
+                BufferedWriter writer = new BufferedWriter(outputStreamWriter);
+                try {
```

Comment:
maybe put the writer as well into try() to have auto-close it?

Reply:
Hi @pierrehenri-dauvergne, I hope you are doing fine. Sorry, I should have been more clear above.
I think it would be worth to use a try-with-resources statement here?